packages/openapi3-parser/lib/parser/oas/parseSchemaObject.js

```diff
@@ -121,13 +125,71 @@ function parseType(context) {
 
   const ensureValidType = R.unless(
     element => isValidType(element.toValue()),
-    createWarning(context.namespace, `'${name}' 'type' must be either ${types.join(', ')}`)
+    createWarning(namespace, `'${name}' 'type' must be either ${types.join(', ')}`)
   );
 
-  return pipeParseResult(context.namespace,
-    R.unless(isString, value => createWarning(context.namespace, `'${name}' 'type' is not a string`, value)),
+  const parseStringType = pipeParseResult(namespace,
     ensureValidType,
-    type => new context.namespace.elements.Array([type]));
+    type => new namespace.elements.Array([type]));
+
+  const ensureValidTypeInArray = R.unless(
+    element => isValidType(element.toValue()),
+    createWarning(namespace, `'${name}' 'type' array must only contain values: ${types.join(', ')}`)
+  );
+
+  const parseArrayTypeItem = pipeParseResult(namespace,
+    R.unless(isString, createWarning(namespace, `'${name}' 'type' array value is not a string`)),
+    ensureValidTypeInArray);
+
+  const isEmpty = arrayElement => arrayElement.isEmpty;
+
+  // ArrayElement -> ParseResult<ArrayElement, Annotation>
+  const ensureTypesAreUnique = (types) => {
```

Comment:
I know this is a bit complicated, I am pretty sure I can make this function a bit simpler tomorrow with fresh eyes.

Reply:
What about using `R.uniqBy`?

Reply:
```js
const types = ['string', 'boolean', 'boolean', 'annotation'].map(t => ({ toValue: () => t }));

const toValue = R.invoker(0, 'toValue');
R.pipe(
  R.uniqBy(toValue),
  R.map(toValue)
)(types)
```

Reply:
Obviously the above doesn't help collecting the Annotations separately. Perhaps you could get that single list along with annotations and then filter them out in the last step:

```js
const validated = R.map(validateType, types);
const annotation = R.find(is(Annotation), validated);
const validatedTypes = R.filter(complement(equals(annotation)), validated);
return new namespace.elements.ParseResult(validatedTypes, annotation);
```